fs.read into zero buffer should not throw exception

feross opened the pull request:
Fixes #4517.

feross commented on src/node_file.cc:
```diff
@@ -1098,7 +1098,7 @@ static void Read(const FunctionCallbackInfo<Value>& args) {
 
   size_t off = args[2]->Int32Value();
-  if (off >= buffer_length) {
+  if (off > buffer_length) {
```
This should be safe because the following if-statement will catch if offset + length would be out-of-bounds of the buffer.

bnoordhuis replied:
I don't think it is.  If `buffer_length == 0`, then it's possible for `buffer_data == nullptr`.  `nullptr + 0` is technically defined behavior (in C++, not C) but passing that pointer onto the kernel can - but does not have to - fail.

If zero-sized reads are desirable, I would handle them in `lib/fs.js`.

feross replied:
Done.

Trott commented:
CI: https://ci.nodejs.org/job/node-test-commit/1611/

thefourtheye commented on test/parallel/test-fs-read-buffer-zero-length.js:
```diff
@@ -0,0 +1,26 @@
+    bufferSync = new Buffer(0),
+    readCalled = 0;
+
+fs.read(fd, bufferAsync, 0, 0, 0, function(err, bytesRead) {
```
Use `common.mustCall` instead of using `readCalled`

feross replied:
I just copied this from another test. Will update.

thefourtheye commented:
What could be the possible use-case which would need this?

Trott commented:
@thefourtheye https://github.com/feross/webtorrent/issues/461

feross commented:
@thefourtheye Yep, what @Trott said.

Zero-length files are valid files. If the user passes one into our module, I think it's reasonable to expect `fs.read` to work. Instead, we currently have to resort to detecting this special case like this: https://github.com/mafintosh/random-access-file/commit/32f3d1b49f0844db50a894502711ea37cef52f86

feross commented:
@thefourtheye Just addressed your comment by updating the PR.

tejasmanohar commented:
:+1:

Trott commented:
Re-running CI since force push: https://ci.nodejs.org/job/node-test-commit/1612/

feross commented:
I just force pushed again to address @bnoordhuis's comments. I moved the logic to `lib/fs.js`, as he suggested.

@Trott You might want to re-run the CI again ;)

Trott commented:
CI round three: https://ci.nodejs.org/job/node-test-pull-request/1136/

evanlucas commented:
CI looks good. LGTM

thefourtheye commented on lib/fs.js:
```diff
@@ -609,6 +609,12 @@ fs.read = function(fd, buffer, offset, length, position, callback) {
   }
 
+  if (length === 0) {
+    return process.nextTick(function() {
```
Hmmm, this is not right. There is actually a deprecated `fs.read` form which returns `string`.

So,

1. This will actually return a buffer and the deprecated `fs.read` will collapse.
2. The order of parameters in string form is actually string and the bytesRead.

feross replied:
Actually, this calls the redefined `callback` function (that starts on line 600), so the right arguments should be passed to the callback function, namely: `null`, `''`, and `0`.

Unless I'm missing something.

feross replied:
You comment prompted me to check the `fs.readSync` case and the issue you describe exists there. I'll fix it and add a test.

feross commented:
Caught one additional bug in the legacy usage of `fs.readSync`, so I fixed it and added a test.

Please take a look at the changes and re-run CI. :+1:

thefourtheye commented on test/parallel/test-fs-read-zero-length.js:
```diff
@@ -0,0 +1,18 @@
+    fd = fs.openSync(filepath, 'r'),
+    expected = '';
+
+fs.read(fd, 0, 0, 'utf-8', common.mustCall(function(err, str, bytesRead) {
```
Make length consistent with this and the `readSync` case.

feross replied:
Done!

thefourtheye commented on lib/fs.js:
```diff
@@ -648,6 +654,14 @@ fs.readSync = function(fd, buffer, offset, length, position) {
     offset = 0;
   }
 
+  if (length === 0) {
```
The check here means that even if the file doesn't exist/inaccessible, if the length is zero, we will return valid values.

feross replied:
How do you propose to solve this? @bnoordhuis asked me to move the checks to `lib/fs.js` to avoid some kernel edge case.

Thoughts?

trevnorris replied:
@thefourtheye Since the fd itself is passed best we can attempt to do is read directly from it. But as @feross mentioned, there are cross platform issues that prevents this being done correctly. I don't think that assuming the fd is alright is a problem in this case.

thefourtheye replied:
@trevnorris Ah, you are right. I totally missed that. Then this will not be a problem here.

jasnell commented:
Defensively marking as `semver-major` due to the error handling change. Minor might be more appropriate.

feross commented:
Updated PR to make the test more consistent, per @thefourtheye's comment.

I think this is ready to be landed :+1:

thefourtheye commented:
CI Run: https://ci.nodejs.org/job/node-test-pull-request/1149/

jasnell commented on test/parallel/test-fs-read-buffer-zero-length.js:
```diff
@@ -0,0 +1,19 @@
+'use strict';
+var common = require('../common');
+var assert = require('assert');
+var path = require('path'),
```
these should use `const`

feross replied:
You guys should really add an `eslint` rule for this. So much overhead for contributors to deal with the back-and-forth.

Fishrock123 replied:
As far as I know, the eslint rule only catches `const` vs `let`, not vs `var`..

cc @silverwind?

jasnell commented on test/parallel/test-fs-read-buffer-zero-length.js:
```diff
@@ -0,0 +1,19 @@
+    fd = fs.openSync(filepath, 'r'),
+    bufferAsync = new Buffer(0),
+    bufferSync = new Buffer(0);
+
```
Nit: would prefer the decl's to be separated but not critical

feross replied:
This was copied from another test file. Will change.

jasnell commented on test/parallel/test-fs-read-zero-length.js:
```diff
@@ -0,0 +1,18 @@
+var common = require('../common');
+var assert = require('assert');
+var path = require('path'),
+    fs = require('fs'),
```
`const` please :-)

feross replied:
Yep.

feross commented:
Force pushed again to address @jasnell's comments.

*fingers crossed*

jasnell commented:
thank you @feross ... and thank you for bearing with us ;-) the contribution is very appreciated. 
Doing another CI pass: https://ci.nodejs.org/job/node-test-pull-request/1151/ 
LGTM if CI is green! :-)

feross commented:
Thanks @jasnell, I appreciate the attention to detail.

jasnell commented:
CI failures look unrelated! Will get this landed

jasnell commented:
Landed in 2b15e68

feross commented:
Nice!
